File: backend/app/schemas/teacher.py
"""
Pydantic schemas for Teacher management.
"""

import re
from typing import Optional, List
from datetime import datetime
from pydantic import BaseModel, EmailStr, Field, ConfigDict, field_validator, model_validator

from app.schemas.user import UserResponseSchema
from app.schemas.goso import SubjectBrief

# Kazakhstan phone format: +7XXXXXXXXXX
_KZ_PHONE_RE = re.compile(r"^\+7\d{10}$")


class TeacherCreate(BaseModel):
    """Schema for creating a new teacher (includes user data)."""

    # User fields — at least one of email/phone must be provided
    email: Optional[EmailStr] = Field(None, description="Teacher email address")
    password: str = Field(..., min_length=8, description="Teacher password (min 8 characters)")
    first_name: str = Field(..., min_length=1, max_length=100, description="First name")
    last_name: str = Field(..., min_length=1, max_length=100, description="Last name")
    middle_name: Optional[str] = Field(None, max_length=100, description="Middle name")
    phone: Optional[str] = Field(None, max_length=50, description="Phone number (+7XXXXXXXXXX)")

    # Teacher fields
    teacher_code: Optional[str] = Field(
        None,
        max_length=50,
        description="Teacher code (auto-generated if not provided)",
    )
    subject_id: Optional[int] = Field(None, description="Subject ID (legacy, single subject)")
    subject_ids: Optional[List[int]] = Field(None, description="List of subject IDs (multi-subject)")
    bio: Optional[str] = Field(None, description="Teacher biography")

    @field_validator("phone")
    @classmethod
    def validate_phone(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v != "":
            if not _KZ_PHONE_RE.match(v):
                raise ValueError("Phone must be in +7XXXXXXXXXX format")
        return v or None

    @model_validator(mode='after')
    def check_email_or_phone(self):
        if not self.email and not self.phone:
            raise ValueError("Either email or phone must be provided")
        return self


class TeacherUpdate(BaseModel):
    """Schema for updating an existing teacher."""

    # Only teacher-specific fields can be updated
    # User fields are updated via UserUpdate schema
    teacher_code: Optional[str] = Field(None, max_length=50, description="Teacher code")
    subject_id: Optional[int] = Field(None, description="Subject ID (legacy, single subject)")
    subject_ids: Optional[List[int]] = Field(None, description="List of subject IDs (multi-subject)")
    bio: Optional[str] = Field(None, description="Teacher biography")


class TeacherResponse(BaseModel):
    """Schema for teacher response (detailed with user data)."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    school_id: int
    user_id: int
    teacher_code: str
    subject_id: Optional[int] = Field(None, description="Subject ID (legacy, first subject)")
    subject: Optional[str] = Field(None, description="Subject name (for backward compatibility)")
    subject_rel: Optional[SubjectBrief] = Field(None, description="Subject details (legacy)")
    subject_ids: List[int] = Field(default_factory=list, description="All subject IDs")
    subjects: List[SubjectBrief] = Field(default_factory=list, description="All subjects")
    bio: Optional[str]
    created_at: datetime
    updated_at: datetime

    # Nested user data
    user: UserResponseSchema

    # Nested classes data (optional, loaded when needed)
    classes: Optional[List["SchoolClassBriefResponse"]] = None


class TeacherListResponse(BaseModel):
    """Schema for teacher list response (brief)."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    teacher_code: str
    subject_id: Optional[int] = Field(None, description="Subject ID (legacy, first subject)")
    subject: Optional[str] = Field(None, description="Subject name (for backward compatibility)")
    subject_rel: Optional[SubjectBrief] = Field(None, description="Subject details (legacy)")
    subject_ids: List[int] = Field(default_factory=list, description="All subject IDs")
    subjects: List[SubjectBrief] = Field(default_factory=list, description="All subjects")
    created_at: datetime

    # Nested user data (brief)
    user: UserResponseSchema


# Import after class definition to avoid circular import
from app.schemas.student import SchoolClassBriefResponse

TeacherResponse.model_rebuild()

File: admin-v2/src/lib/validations/teacher.ts
import { z } from 'zod';

const KZ_PHONE_RE = /^\+7\d{10}$/;

// Schema for creating a new teacher
export const teacherCreateSchema = z.object({
  email: z
    .string()
    .email('Неверный формат email')
    .optional()
    .or(z.literal('')),
  password: z
    .string()
    .min(8, 'Минимум 8 символов')
    .max(100, 'Максимум 100 символов'),
  first_name: z
    .string()
    .min(1, 'Имя обязательно')
    .max(100, 'Максимум 100 символов'),
  last_name: z
    .string()
    .min(1, 'Фамилия обязательна')
    .max(100, 'Максимум 100 символов'),
  middle_name: z
    .string()
    .max(100, 'Максимум 100 символов')
    .optional()
    .or(z.literal('')),
  phone: z
    .string()
    .max(50, 'Максимум 50 символов')
    .optional()
    .or(z.literal('')),
  teacher_code: z
    .string()
    .max(50, 'Максимум 50 символов')
    .optional()
    .or(z.literal('')),
  subject_id: z
    .number()
    .int()
    .positive()
    .optional()
    .nullable(),
  bio: z
    .string()
    .max(1000, 'Максимум 1000 символов')
    .optional()
    .or(z.literal('')),
}).refine(
  (data) => {
    const hasEmail = !!data.email && data.email.length > 0;
    const hasPhone = !!data.phone && KZ_PHONE_RE.test(data.phone);
    return hasEmail || hasPhone;
  },
  {
    message: 'Укажите email или телефон в формате +7XXXXXXXXXX',
    path: ['phone'],
  }
);

// Schema for updating a teacher (without email/password)
export const teacherUpdateSchema = z.object({
  first_name: z
    .string()
    .min(1, 'Имя обязательно')
    .max(100, 'Максимум 100 символов')
    .optional(),
  last_name: z
    .string()
    .min(1, 'Фамилия обязательна')
    .max(100, 'Максимум 100 символов')
    .optional(),
  middle_name: z
    .string()
    .max(100, 'Максимум 100 символов')
    .optional()
    .or(z.literal('')),
  phone: z
    .string()
    .max(50, 'Максимум 50 символов')
    .optional()
    .or(z.literal('')),
  teacher_code: z
    .string()
    .max(50, 'Максимум 50 символов')
    .optional()
    .or(z.literal('')),
  subject_id: z
    .number()
    .int()
    .positive()
    .optional()
    .nullable(),
  bio: z
    .string()
    .max(1000, 'Максимум 1000 символов')
    .optional()
    .or(z.literal('')),
});

// Infer types from schemas
export type TeacherCreateInput = z.infer<typeof teacherCreateSchema>;
export type TeacherUpdateInput = z.infer<typeof teacherUpdateSchema>;

// Default values for forms
export const teacherCreateDefaults: TeacherCreateInput = {
  email: '',
  password: '',
  first_name: '',
  last_name: '',
  middle_name: '',
  phone: '',
  teacher_code: '',
  subject_id: null,
  bio: '',
};
